### PR TITLE
Added backend models and functions for Reports

### PR DESCRIPTION
**Pull request by darbrius**

### CODE REVIEW

File: server/app.js [hunk 1]
```diff
@@ -8,6 +8,9 @@ const DataAccessLayer = require('./controllers/DataAccessLayer.js')
 
 http.listen(3000, () => {
     let users = DataAccessLayer.ReadUsersFile();
+    console.log(users);
+    let reports = DataAccessLayer.ReadReportsFile();
+    console.log(reports);
```
Comment (rzayevan) [inline server/app.js]: Don't think you should log these. Maybe comment them out?

File: server/controllers/DataAccessLayer.js [hunk 2]
```diff
@@ -102,6 +110,145 @@ exports.UpdateUser = function(user) {
...
+
+    // Create a regex that will be searched for within the text file
+    let originalReportStringRegex = ".*" + originalReport.id + ".*";
+
+    // Create a new Date object to set the lastUpdatedDate's to
+    let newLastUpdatedDate = new Date();
+
+    let newChatLogString = GetChatLogString(report.chatLogs);
+
+    // Create a new string to update the file to
```
Comment (rzayevan) [inline server/controllers/DataAccessLayer.js]: Don't think comments on lines 203 and 208 are complete

File: server/controllers/DataAccessLayer.js [hunk 3]
```diff
@@ -102,6 +110,145 @@ exports.UpdateUser = function(user) {
...
+        files: "data/Reports.txt",
+        from: regex,
+        to: newReportString
+    }
+
+    // Replace the relevant contents of the text file (replace-in-file node package)
+    replace(replaceOptions)
+        .then(result => {
+            // If no error occurred and the result states that the file was successfully changed
+            if (result.hasChanged === true) {
```
Comment (rzayevan) [inline server/controllers/DataAccessLayer.js]: Is the `===` check necessary?

File: server/controllers/DataAccessLayer.js [hunk 4]
```diff
@@ -102,6 +110,145 @@ exports.UpdateUser = function(user) {
...
+ * Returns a string based on the supplied array of ChatMessage's
+ * @param {ChatMessage[]} chatMessages An array of ChatMessage's that is supplied
+ */
+GetChatLogString = function(chatMessages) {
+    let chatLogString = "";
+    // For each ChatMessage in the given Report's chatLog
+    for (let i = 0; i < chatMessages.length; i++) {
+        // Create a unique string for the message
+        let chatMessageString = "{" + chatMessages[i].username + "," + chatMessages[i].message + "}";
+        // Apoend the string to the chatLogString
```
Comment (rzayevan) [inline server/controllers/DataAccessLayer.js]: typo

File: server/controllers/DataAccessLayer.js [hunk 5]
```diff
@@ -102,6 +110,145 @@ exports.UpdateUser = function(user) {
...
+        })
+        .catch(err => {
+            // Log the error if the text file is not successfully updated
+            console.log(err);
+    });
+}
+
+/**
+ * Returns a string based on the supplied array of ChatMessage's
+ * @param {ChatMessage[]} chatMessages An array of ChatMessage's that is supplied
```
Comment (rzayevan) [inline server/controllers/DataAccessLayer.js]: `'s` is extra

**Review by rzayevan — APPROVED** [review]

File: server/controllers/DataAccessLayer.js [hunk 6]
```diff
@@ -1,9 +1,27 @@
...
+ * Retrieves the array of cached Reports
+ */
+exports.getCachedReports = function() {
+    return cachedReports;
+}
+
+/**
+ * Retrieves cached users
+ */
+exports.GetCachedUsers = function() {
```
Comment (rzayevan) [inline server/controllers/DataAccessLayer.js:22]: I think I've added a similar function that is pushed to master

File: server/controllers/DataAccessLayer.js (hunk 3 above)
Comment (darbrius) [reply server/controllers/DataAccessLayer.js]: Yes, it allows for verification that the file was successfully updated with errors, then I update the cache

File: server/controllers/DataAccessLayer.js (hunk 6 above)
Comment (darbrius) [reply server/controllers/DataAccessLayer.js:22]: This is the master one, I moved it.

Comment (rzayevan) [reply server/controllers/DataAccessLayer.js:22]: Nevermind, I saw that you just moved it to the top

File: server/controllers/DataAccessLayer.js (hunk 5 above)
Comment (darbrius) [reply server/controllers/DataAccessLayer.js]: Fixed

File: server/controllers/DataAccessLayer.js (hunk 4 above)
Comment (darbrius) [reply server/controllers/DataAccessLayer.js]: Fixed

File: server/controllers/DataAccessLayer.js (hunk 3 above)
Comment (rzayevan) [reply server/controllers/DataAccessLayer.js]: Can't you just do `if (result.hasChanged)`?

File: server/controllers/DataAccessLayer.js (hunk 5 above)
Comment (rzayevan) [reply server/controllers/DataAccessLayer.js]: Is it? 😄 (line 251)

Comment (darbrius) [reply server/controllers/DataAccessLayer.js]: Yes...

Comment (rzayevan) [reply server/controllers/DataAccessLayer.js]: Now it is 😇

**Pull request merged**